Backend

pyukey opened the pull request:
Set up backend so you can now do:
- API calls for users, scooters, and transactions
- User login/out/signup

Maybe should test security before merging

pyukey committed:
Setup Database

pyukey committed:
API setup

Example uses are commented in src/routes/+page.svelte

pyukey committed:
Users Exist!

Nikhil-42 commented on package.json:
```diff
@@ -44,8 +46,11 @@
 		"drizzle-orm": "^0.31.2",
+		"lucia": "^3.2.0",
+		"pg": "^8.12.0",
 		"postgres": "^3.4.4"
```
If you switched to `pg`, please remove `postgres`.

Nikhil-42 commented on src/lib/db/index.ts:
```diff
@@ -0,0 +1,13 @@
+
+export const client = new pg.Client({
+  user: 'postgres',
+  password: '07kjhsuka',
```
Use a dotenv variable for sensitive information.

pyukey replied:
Oops! thank you for catching this

Nikhil-42 commented on src/lib/db/index.ts:
```diff
@@ -0,0 +1,13 @@
+  password: '07kjhsuka',
+  host: 'localhost',
+  port: 5432,
+  database: 'postgres',
```
Tbh these should all be dotenv vars.

Nikhil-42 commented on src/lib/db/migrate.ts:
We are using drizzle-kit, a custom migrate script should not be necessary.

pyukey replied:
Migrations were not working for me earlier. Feel free to change.

Nikhil-42 commented on src/lib/db/schema.ts:
```diff
@@ -0,0 +1,75 @@
+	}).notNull()
+});
+
+export const users = pgTable("users", {
+  id: text("id").primaryKey().notNull(),
+  firstname: text("firstname").notNull(),
+  lastname: text("lastname").notNull(),
+  email: text("email").notNull(),
+  passwordHash: text("password_hash").notNull(),
+  credit: integer("credit"),
+  isAdmin: boolean("is_admin")
+});
```
Remember the discussion we had earlier about including Customer and Employee tables so user can be reserved for login only

pyukey replied:
Yes, and in this conversation I decided against that since it would make querying more complicated, and require multiple queries.

Nikhil-42 replied:
It makes it much easier and safer to define relations. For example, only someone in the employees table should be able to approve a rental.

Nikhil-42 commented on src/lib/db/schema.ts:
```diff
@@ -0,0 +1,75 @@
+  amount: integer("amount"),
+  checkInTime: date("check_in_time"),
+  checkOutTime: date("check_out_time")
+})
```
Similarly, transactions don't always include a scooter (ex. Employee directly adds $50 to an customers's credit)

Should have a transactions table to track adding/removing credit, and a separate rentals table with a one-to-one relation to transactions.

pyukey replied:
That is a good point. Adding a separate table sounds like the way to go. In the meantime I will just make scooterId an optional parameter.

Nikhil-42 commented on src/lib/db/schema.ts:
```diff
@@ -0,0 +1,75 @@
+
+export const transactions = pgTable("transactions", {
+  id: serial("id").primaryKey().notNull(),
+  customerId: text("customer_id").references(() => users.id, {onDelete: "cascade"}).notNull(),
+  scooterId: text("scooter_id").references(() => scooters.id, { onDelete: "cascade" }).notNull(),
+  employeeId: text("employee_id").references(() => users.id, {onDelete: "cascade"}).notNull(),
```
Imo we should never delete transactions.

pyukey replied:
Ideally yes, but it's good to have the feature in place in case of accidents.

Nikhil-42 replied:
I'd rather have an error than allow silently deleting history.

Nikhil-42 commented on src/lib/server/auth.ts:
```diff
@@ -0,0 +1,36 @@
+import { DrizzlePostgreSQLAdapter } from "@lucia-auth/adapter-drizzle";
+import { sessionTable, users } from "../db/schema";
```
Use $lib instead of relative paths.

pyukey replied:
Good point

Nikhil-42 commented on src/lib/server/auth.ts:
```diff
@@ -0,0 +1,36 @@
+import { DrizzlePostgreSQLAdapter } from "@lucia-auth/adapter-drizzle";
+import { sessionTable, users } from "../db/schema";
+import db from "../db";
```
See above

pyukey committed:
BACKEND COMPLETE

Added user authentication
Fixed API calls
Added helper functions in calls.ts
Included examples inn +page.svelte

Nikhil-42 commented on src/app.d.ts:
```diff
@@ -3,7 +3,10 @@
 		// interface Error {}
-		// interface Locals {}
+		interface Locals {
+			user: import("lucia").User | null;
+			session: import("lucia").Session | null;
```
Switch to ES6 modules import if no side-effects

pyukey replied:
You can change it if you want. This is what the Lucia documentation recommends and I'd rather not mess with things I don't understand: https://lucia-auth.com/getting-started/sveltekit

Nikhil-42 commented on src/lib/calls.ts:
```diff
@@ -0,0 +1,232 @@
+        return { success: false };
+    }
+}
+export async function updateScooter(scooterId: string, latitude: number, longitude: number, battery: number, needRepairs: boolean, checkedOut: boolean) {
```
For an update, everything but id should be optional

Jaidenmagnan replied:
Possibly add functions to updateScooter to easily change checkedOut and needRepairs. Also add checked out and need repairs to add scooter. Mostly looks great though.

Nikhil-42 replied:
I think we assume newly added scooters are in good condition and not checked out.

Nikhil-42 commented on src/lib/calls.ts:
```diff
@@ -0,0 +1,232 @@
+        return { success: false };
+    }
+}
+export async function getScooter() {
```
Should this be plural?

pyukey replied:
sure

Nikhil-42 commented on src/lib/calls.ts:
```diff
@@ -0,0 +1,232 @@
+        const response = await fetch(APIT, {
+            method: 'DELETE',
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({ transactionId })
+        });
+
+        if (!response.ok) {
+            throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+
+        const data = await response.json();
+        return data;
+    } catch (error) {
+        console.error('Error fetching scooters:', error);
+        return { success: false };
+    }
+}
+export async function updateTransaction(amount: number, transactionId: string) {
+    try {
+        const response = await fetch(APIT, {
+            method: 'PATCH',
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({ amount, transactionId })
+        });
+
+        if (!response.ok) {
+            throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+
+        const data = await response.json();
+        return data;
+    } catch (error) {
+        console.error('Error fetching scooters:', error);
+        return { success: false };
+    }
+}
```
Should not update or delete for transactions

pyukey replied:
Yeah I agree. Having an immutable transactions table and a mutable rentables table sounds smart. If there is ever an error in the transaction history, we can always issue a refund as a later transaction.

Nikhil-42 commented on src/lib/calls.ts:
Please add types to any exported methods

pyukey replied:
Yeah... will take some time to update them all

Nikhil-42 commented on src/lib/calls.ts:
```diff
@@ -0,0 +1,232 @@
+        const data = await response.json();
+        return data;
+    } catch (error) {
+        console.error('Error fetching scooters:', error);
```
Copy paste issue

pyukey replied:
?

Nikhil-42 replied:
The error message says "Error fetching scooters"

Nikhil-42 commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,20 @@
+import { lucia } from "$lib/server/auth";
+import { fail, redirect, type Actions, type ServerLoad} from "@sveltejs/kit";
+export const load: ServerLoad = async ({ locals }) => {
+	// ...
+};
```
What does this do?

pyukey replied:
Nothing! Sorry I was just following the tutorial (go to bottom): https://lucia-auth.com/tutorials/username-and-password/sveltekit

Nikhil-42 commented on src/lib/db/schema.ts:
```diff
@@ -0,0 +1,75 @@
+})
+
+// Define relationships for transactions
+export const transactionsRelations = relations(transactions, ({ one }) => ({
+  customer: one(users, {
+    fields: [transactions.customerId],
+    references: [users.id],
+  }),
+  scooter: one(scooters, {
+    fields: [transactions.scooterId],
+    references: [scooters.id],
+  }),
+  employee: one(users, {
+    fields: [transactions.employeeId],
+    references: [users.id],
+  }),
+}));
+
+// Define relationships for users
+export const usersRelations = relations(users, ({ many }) => ({
+  transactionsAsCustomer: many(transactions)
+}))
+
+// Define relationships for scooters
+export const scootersRelations = relations(scooters, ({ many }) => ({
+  transactions: many(transactions),
+}));
+
+export const sessionRelations = relations(sessionTable, ({ one }) => ({
+  user: one(users, {
+    fields: [sessionTable.userId],
+    references: [users.id],
+  }),
```
In theory, we shouldn't need these if we already added FOREIGN key constraints with .references

pyukey replied:
I believe it's good practice in Drizzle to also explicity declare relations like this. I think it makes querying more efficient somehow

Nikhil-42 replied:
Makes sense

Nikhil-42 commented on drizzle.config.ts:
```diff
@@ -3,7 +3,7 @@ import { defineConfig } from "drizzle-kit";
 export default defineConfig({
   dialect: "postgresql",
-  schema: "./src/lib/schema.ts",
+  schema: "./src/lib/db/schema.ts",
```
Need to select the `pg` driver

pyukey replied:
Tbh I'm not sure how to do this correctly, or if it is necessary

Nikhil-42 commented on src/lib/db/schema.ts:
```diff
@@ -0,0 +1,75 @@
+  email: text("email").notNull(),
+  passwordHash: text("password_hash").notNull(),
+  credit: integer("credit"),
+  isAdmin: boolean("is_admin")
```
Replace with a text role that can be either "customer", "employee", or "admin".

pyukey replied:
Why? booleans can store 3 values (null, false, true), take up less memory, and are less prone to errors from typoes

Nikhil-42 commented on src/lib/server/auth.ts:
```diff
@@ -0,0 +1,36 @@
+
+interface DatabaseUserAttributes {
+	userId: string
+  firstname: string
+  lastname: string
```
Fix indentation

Nikhil-42 commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,20 @@
+		event.cookies.set(sessionCookie.name, sessionCookie.value, {
+			path: ".",
+			...sessionCookie.attributes
+		});
```
Don't forget to clear `event.locals.user` and `event.locals.session`

pyukey replied:
I believe that is what invalidateSession does?

Nikhil-42 replied:
lucia.invalidateSession only updates the DB not the event.locals. It doesn't really matter because we are redirecting anyways (which clears the event.locals) but it's good practice.

Nikhil-42 commented on src/routes/+page.svelte:
```diff
@@ -1,7 +1,27 @@
 <script>
+	import { enhance } from "$app/forms";
   import "../app.css";
```
Limit use of relative imports

Nikhil-42 commented on src/routes/+page.svelte:
```diff
@@ -15,6 +35,10 @@
 		try editing <strong>src/routes/+page.svelte</strong>
 	</h2>
+	
+	<form method="post" use:enhance>
```
```suggestion
	<form method="POST" use:enhance>
```

Nikhil-42 commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,20 @@
+	// ...
+};
+
+export const actions: Actions = {
```
Please add a comment explaining this is attached to the Sign Out form.

Nikhil-42 commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,20 @@
+export const actions: Actions = {
+	default: async (event) => {
+		if (!event.locals.session) {
+			return fail(401);
```
Instead of failing, redirect to login.

pyukey replied:
Sounds good

Nikhil-42 commented on src/routes/api/scooters/+server.ts:
Tbh we shouldn't need manual api and fetch definitions. We should handle all data hydration with `load`.

pyukey replied:
idk how to do that

Nikhil-42 commented on src/routes/login/+page.server.ts:
```diff
@@ -0,0 +1,67 @@
+		}
+
+		const existingUser = await db.query.users.findFirst({
+			where: (table, { eq }) => {return eq(table.firstname, username)}
```
The username cannot be firstname use email instead

pyukey replied:
Good catch!

Nikhil-42 commented on src/routes/signup/+page.server.ts:
```diff
@@ -0,0 +1,59 @@
+		await db.insert(users).values({
+			id: userId,
+			firstname: username,
+            lastname: username,
+            email: username,
```
Fix indentation

pyukey replied:
I'm not suer what you mean. If you don't like how it's formatted you can change it.

Nikhil-42 commented on src/routes/signup/+page.server.ts:
```diff
@@ -0,0 +1,59 @@
+		// TODO: check if username is already used
+		await db.insert(users).values({
+			id: userId,
+			firstname: username,
```
Add these entries to the form, don't copy-paste username into everything

pyukey replied:
Yep, absolutely. I was just testing to see if it works so I didn't add the other fields

Nikhil-42 commented:
@pyukey Great work, left a few comments.

pyukey committed:
Changed relative pathing

pyukey committed:
Remove unneccesarry code

pyukey committed:
Changed parameters for signup/login

Nikhil-42 committed:
Update package-lock.json

Nikhil-42 committed:
Switch to ES6 module import

Nikhil-42 committed:
Use environment for database connection

Nikhil-42 committed:
Fix relative imports

Nikhil-42 committed:
Update database schema

Nikhil-42 committed:
Make sure to clear locals after log out

Nikhil-42 committed:
Fix prerendering failure

Nikhil-42 committed:
Perform migration

Nikhil-42 committed:
Move protected routes to (authed) group

Nikhil-42 committed:
Document hooks

Nikhil-42 committed:
Create store for user info

Nikhil-42 committed:
Migrate Header and Footer

Nikhil-42 committed:
Update schema with pgView

Nikhil-42 committed:
Inject balance, isAdmin, and role

Nikhil-42 committed:
Protect (authed) routes

Nikhil-42 committed:
Remove duplicate session resolutions

Nikhil-42 committed:
Use proper http error codes

Nikhil-42 committed:
Fix impossible to create employees

Nikhil-42 committed:
Remove raw sql query usage

Nikhil-42 commented on src/lib/stores.ts:
```diff
@@ -0,0 +1,4 @@
+import { type User } from "lucia";
+import { writable } from "svelte/store";
+
+export const user = writable<User | null>(null);
```
Store is unused, populate in layout.svelte

Nikhil-42 reviewed (commented):
Didn't review API, trial by fire more efficient

Nikhil-42 committed:
Actualizar +page.svelte

Jaidenmagnan commented on src/lib/calls.ts:
```diff
@@ -0,0 +1,232 @@
+        return { success: false };
+    }
+}
+export async function addTransaction(customerId: string, scooterId: string) {
```
Should this automatically set checkedOut to true?

Jaidenmagnan replied:
Do we also store failed transactions.

Nikhil-42 replied:
I separated transactions from rentals

Jaidenmagnan reviewed (approved):
I added comments throughout, mostly to API function calls, just quality of life changes though. Everything looks good.

Nikhil-42 reviewed (commented):
> Do we also store failed transactions. 

Transactions represent a change in a users balance, I moved rentals to a different table.

Nikhil-42 committed:
Clean up calls.ts

Nikhil-42 committed:
Merge remote-tracking branch 'origin/main' into backend
